plugins/trip.hpp

```diff
@@ -0,0 +1,332 @@
+                !route_parameters.hints[i].empty()) {
+                PhantomNode current_phantom_node;
+                ObjectEncoder::DecodeFromBase64(route_parameters.hints[i], current_phantom_node);
+                if (current_phantom_node.is_valid(facade->GetNumberOfNodes()))
+                {
+                    phantom_node_vector[i].emplace_back(std::move(current_phantom_node));
+                    continue;
+                }
+            }
+            facade->IncrementalFindPhantomNodeForCoordinate(route_parameters.coordinates[i],
+                                                            phantom_node_vector[i], 1);
+            if (phantom_node_vector[i].size() > 1) {
+                phantom_node_vector[i].erase(std::begin(phantom_node_vector[i]));
+            }
+            BOOST_ASSERT(phantom_node_vector[i].front().is_valid(facade->GetNumberOfNodes()));
+        }
+    }
+
+    // Object to hold all strongly connected components (scc) of a graph
+    // to access all graphs with component ID i, get the iterators by:
+    // auto start = std::begin(scc_component.component) + scc_component.range[i];
+    // auto end = std::begin(scc_component.component) + scc_component.range[i+1];
+    struct SCC_Component{
+        // in_component: all NodeIDs sorted by component ID
+        // in_range: index where a new component starts
+        //
+        // example: NodeID 0, 1, 2, 4, 5 are in component 0
+        //          NodeID 3, 6, 7, 8    are in component 1
+        //          => in_component = [0, 1, 2, 4, 5, 3, 6, 7, 8]
+        //          => in_range = [0, 5]
+        SCC_Component(std::vector<NodeID> in_component,
+                      std::vector<size_t> in_range)
+                      : component(in_component),
+                        range(in_range) {
+            range.push_back(in_component.size());
+            BOOST_ASSERT_MSG(in_component.size() >= in_range.size(),
+                             "scc component and its ranges do not match");
+            BOOST_ASSERT_MSG(*std::max_element(in_range.begin(), in_range.end()) < in_component.size(),
+                             "scc component ranges are out of bound");
+            BOOST_ASSERT_MSG(*std::min_element(in_range.begin(), in_range.end()) >= 0,
+                             "invalid scc component range");
+            BOOST_ASSERT_MSG([&in_range](){
+                                 for (std::size_t r = 0; r < in_range.size() - 1; ++r) {
+                                     if (in_range[r] > in_range[r+1]) {
+                                         return false;
+                                     }
+                                 }
+                                 return true;
+                             }(),
+                             "invalid component ranges");
+          };
+
+        // constructor to use when whole graph is one single scc
+        SCC_Component(std::vector<NodeID> in_component)
```

Comment:
Yes, except that the extra `;` at the end is not needed and you make an extra copy when taking `in_component` by-value (but a move would be problematic here since you need the vector's size). Otherwise this *looks* fine.